### PR TITLE
[WIP] HTLC implementation in the wallet

### PR DESCRIPTION
**Pull request by ebfull**
This adds support for timelocked contracts which can be used to atomically
swap hash preimages over the blockchain. It is very similar to the script
used by the lightning network.

Written in collaboration with Pieter Wuille.

### CODE REVIEW

**Comment by jonasschnelli** [comment]
Interesting!
Haven't looked at it in detail, but would this affect "policy" by adding the `TX_HTLC` as to the standard templates?

**Comment by sipa** [comment]
@jonasschnelli All P2SH is already standard.

**Comment by jonasschnelli** [comment]
@sipa: Ah. Right. Then – I guess – the `standard.cpp -> Solver()` changes are only because we want to detect these types of P2SH scripts in the wallet.

**Comment by sipa** [comment]
@jonasschnelli To be clear: this indeed does make the HTLC type transaction standard in non-P2SH setting; it's however still smaller and cheaper than the largest raw multisig that is allowed. Perhaps that indeed shouldn't happen.

**Comment by gmaxwell** [comment]
Concept ACK.

**Comment by dcousens** [comment]
concept ACK

**Comment by petertodd** [comment]
Interestingly, this is useful for Paypub, as well as ZKCP, among other things.

Concept ACK

**Comment by petertodd** [comment]
@sipa For Paypub, having the payments be visible in the blockchain for the receiver is extremely useful; w/o the ability to do so in scriptPubKey you'd likely implement it with an extra OP_RETURN output (maybe until we have a better Bitmessage to use instead). Not necessarily an argument either for or against doing so, but worth noting.

**Comment by btcdrak** [comment]
Concept ACK

**Comment by paveljanik** [comment]
Some tests fail with:
```
Running testscript wallet.py ...
Initializing test directory /tmp/testr_T8PJ
start_node: bitcoind started, calling bitcoin-cli -rpcwait getblockcount
start_node: calling bitcoin-cli -rpcwait getblockcount returned
JSONRPC error: preimage must be hexadecimal string (not '')
```
I can't reproduce locally 8)

**Comment by paveljanik** [comment]
Concept ACK
Needs some tests. Needs BIP for new `OP_ANYDATA`.

**Comment by jonasschnelli** [comment]
Concept ACK.

1.) Not sure if this requires a BIP, I guess we don't need a BIP for every contract template and I can't find the other template matching params (like `OP_SMALLINTEGER`) in the BIPs. 

I also concept ACK the IsStandard rule for HTLC to allow blockchain-visible HTLC, though, this might require a BIP.

**Comment by sipa** [comment]
OP_ANYDATA is just local wallet logic, it uses a range of script opcodes
that are disabled in execution, and is only used inside the matching
templates (I think it's a bad idea to mix them in the same CScript
type...), so I disagree with needing a BIP for that.

IsStandard changes so far have also been done without a BIP, but maybe we
can postpone making raw HTLC standard for now.

Certainly agree we need tests.

**Comment by luke-jr** [comment]
I agree IsStandard changes do not need a BIP, but it seems any transaction type supported by the reference implementation necessarily involves coordination with other software/people, and therefore should have a BIP written describing how it works.

Is there a benefit to bare HTLC over P2SH (or SegWit) HTLC? If not, it seems pointless to expand the policy to allow for it. Note that many users do not enable bare multisig (and it should probably be disabled by default).

**Comment by ebfull** [comment]
I don't think initially we should make this standard outside P2SH (yet), so that's definitely a change that should be made to this PR.

This work so far allows you to "import" a preimage into CWallet (so that you can redeem the atomic swap branch of the script) but it does not persist this to disk, and I'm curious if others think it should. (I'm skeptical of making bitcoind store secrets, especially if for some use cases they are ephemeral.)

**Comment by gmaxwell** [comment]
As an example of what this was written to enable: https://bitcoincore.org/en/2016/02/26/zero-knowledge-contingent-payments-announcement/

**Comment by jtimon** [comment]
Wow, concept ACK.

**Comment by sipa** [comment]
We should probably also wait until CSV is in, and add support for it in the constructed scripts here.

File: src/rpc/rawtransaction.cpp [hunk 1]
```diff
@@ -839,3 +839,4 @@ UniValue sendrawtransaction(const UniValue& params, bool fHelp)
 
     return hashTx.GetHex();
 }
+
```
Comment (jtimon) [inline src/rpc/rawtransaction.cpp]: This new line?

**Comment by gmaxwell** [comment]
@sipa CSV is almost in, but I don't think we should be deploying enabled generation for them until the softfork is good and settled.  Maybe behind a hidden option or testnet only?

**Comment by ebfull** [comment]
Additionally, supporting RIPEMD_160 for ZKCPs is a must. I will try to find some time to write an arithmetic circuit for it.

**Comment by ebfull** [comment]
After segwit lands (?) I will try to rebase and clean this PR up.

**Comment by btcdrak** [comment]
@ebfull segwit has been merged to master and is also activated on testnet.

**Comment by gmaxwell** [comment]
CSV is active on mainnet!

**Comment by ebfull** [comment]
Rebased this on to master. Now supports RIPEMD160. Switched to CSV (block denominated mode). Added some RPC tests.

It can be trivially extended to use CLTV in the RPC, I just need to come up with a solid UX for it. The same goes for "relative" time strings like "10h" or "5d." Also, still need to bring over the "preimage" display that showed in `listtransactions`.

I'll try to work alongside #7534 for the remaining work, including submitting a BIP.

File: src/script/standard.cpp [hunk 2]
```diff
@@ -54,6 +55,26 @@ bool Solver(const CScript& scriptPubKey, txnouttype& typeRet, vector<vector<unsi
...
+
+        // HTLC where sender requests preimage of a hash
+        {
+            const opcodetype accepted_hashers[] = {OP_SHA256, OP_RIPEMD160};
+            const opcodetype accepted_timeout_ops[] = {OP_CHECKLOCKTIMEVERIFY, OP_CHECKSEQUENCEVERIFY};
+
+            BOOST_FOREACH(opcodetype hasher, accepted_hashers) {
+                BOOST_FOREACH(opcodetype timeout_op, accepted_timeout_ops) {
+                    mTemplates.insert(make_pair(TX_HTLC, CScript()
+                        << hasher << OP_ANYDATA << OP_EQUAL
```
Comment (jl2012) [inline src/script/standard.cpp]: In the "ELSE" case any relay node may replace the top stack with anything without invalidating the tx. See related discussion at https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2016-August/013036.html

**Comment by afk11** [comment]
Needs rebase

**Comment by jtimon** [comment]
Still needs rebase.

**Comment by ebfull** [comment]
Before I rebase, I need to understand the consensus regarding preventing malleability of these scripts.

https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2016-August/013036.html

Multiple suggestions were made and I'm not personally qualified to pick from them.

**Comment by jl2012** [comment]
@ebfull: I'd suggest this one. But you may want to discuss it on mailing list to make sure it is compatible with other implementations (if any)

`OP_IF
 [HASHOP] <digest> OP_EQUALVERIFY
 <seller pubkey>
OP_ELSE
 <num> [TIMEOUTOP] OP_DROP
 <buyer pubkey>
OP_ENDIF
OP_CHECKSIG`

**Comment by da2ce7** [comment]
@ebfull, this looks very interesting to me and many others. Would be wonderful for you can find the time to have a second look at this pull request since v0.14 has been forked off.

**Comment by ebfull** [comment]
I will be rebasing using @jl2012's script as the foundation, and update my BIP draft.

File: src/policy/policy.cpp [hunk 3]
```diff
@@ -46,6 +46,8 @@ bool IsStandard(const CScript& scriptPubKey, txnouttype& whichType, const bool w
             return false;
         if (m < 1 || m > n)
             return false;
+    } else if (whichType == TX_HTLC) {
+        return false;
```
Comment (arielgabizon) [inline src/policy/policy.cpp]: It looks like you're assuming all TX_HTLC transactions are DOS attacks?
Why reject them all?

File: src/script/standard.cpp [hunk 4]
```diff
@@ -374,6 +374,32 @@ CScript GetScriptForMultisig(int nRequired, const std::vector<CPubKey>& keys)
...
 
+CScript GetScriptForHTLC(const CPubKey& seller,
+                         const CPubKey& refund,
+                         const std::vector<unsigned char> image,
+                         uint32_t timeout,
+                         opcodetype hasher_type,
+                         opcodetype timeout_type)
+{
+    CScript script;
+
```
Comment (JeremyRubin) [inline src/script/standard.cpp]: Check that hasher_type and image.size() match?

File: src/rpc/misc.cpp [hunk 5]
```diff
@@ -336,6 +336,89 @@ UniValue createmultisig(const JSONRPCRequest& request)
...
+{
+    if (request.fHelp || request.params.size() != 4) {
+        string msg = "createhtlc seller_key refund_key hash timeout_type timeout\n"
+            "\nCreates an address whose funds can be unlocked with a preimage or as a refund\n"
+            "It returns a json object with the address and redeemScript.\n"
+
+            "\nArguments:\n"
+            "1. seller_key    (string, required) The public key of the possessor of the preimage.\n"
+            "2. refund_key    (string, required) The public key of the recipient of the refund.\n"
+            "3. hash          (string, required) SHA256 or RIPEMD160 hash of the preimage.\n"
```
Comment (JeremyRubin) [inline src/rpc/misc.cpp]: Maybe hash should be optional, and if not supplied a `GetRandHash` preimage should be selected and returned to the caller.

File: src/keystore.h [hunk 6]
```diff
@@ -102,6 +115,16 @@ class CBasicKeyStore : public CKeyStore
     virtual bool HaveCScript(const CScriptID &hash) const;
     virtual bool GetCScript(const CScriptID &hash, CScript& redeemScriptOut) const;
 
+    bool GetPreimage(
+        const std::vector<unsigned char>& image,
+        std::vector<unsigned char>& preimage
+    ) const;
+    
+    bool AddPreimage(
```
Comment (JeremyRubin) [inline src/keystore.h]: I think that you should persist the PreimageMap to disk.

File: src/keystore.h [hunk 7]
```diff
@@ -102,6 +115,16 @@ class CBasicKeyStore : public CKeyStore
...
+        const std::vector<unsigned char>& image,
+        const std::vector<unsigned char>& preimage
+    );
+
```
Comment (JeremyRubin) [inline src/keystore.h]: Maybe add a RemovePreimage, LockPreimage, and UnlockPreimage call, which could be used to ensure that a node no longer has the ability to solve a particular HTLC or does not until explicitly allowed. These should also go to disk.

File: src/wallet/rpcdump.cpp [hunk 8]
```diff
@@ -640,6 +640,49 @@ UniValue dumpwallet(const JSONRPCRequest& request)
...
 
+UniValue importpreimage(const JSONRPCRequest& request)
+{
+    if (!EnsureWalletIsAvailable(request.fHelp))
+        return NullUniValue;
+
+    if (request.fHelp || request.params.size() != 1)
+        throw runtime_error(
+            "importpreimage \"data\"\n"
+            "\nImports a preimage for use in HTLC transactions. Only remains in memory.\n"
```
Comment (JeremyRubin) [inline src/wallet/rpcdump.cpp]: If you add persistence, be sure to update here.

File: src/script/ismine.cpp [hunk 9]
```diff
@@ -125,6 +125,19 @@ isminetype IsMine(const CKeyStore &keystore, const CScript& scriptPubKey, bool&
         break;
     }
 
+    case TX_HTLC:
+    {
+        // Only consider HTLC's "mine" if we own ALL the keys
+        // involved.
+        vector<valtype> keys;
+        keys.push_back(vSolutions[1]);
```
Comment (JeremyRubin) [inline src/script/ismine.cpp]: Maybe just for safety, assert(vSolutions.size() > 3)

**Review by JeremyRubin — COMMENTED** [review]
Concept Ack.

I did some code review, overall looks very clean and concise.

I was slightly concerned with the use of new template parameters, but I think that the proposed new ones are reasonable & general purpose.

I think that the preimage map should persist entries; but perhaps that can be done as a later PR. I would also suggest adding a Lock/Unlock feature in RPC, because I think that that would ensure safer use of entries stored in the preimage map for users.

**Comment by ebfull** [comment]
Thanks for the reviews @arielg and @JeremyRubin. I intend to rebase this again and clean it up.

Also, BIP199 has been submitted to start standardizing what is in this PR.

File: src/script/ismine.cpp [hunk 10]
```diff
@@ -125,6 +125,19 @@ isminetype IsMine(const CKeyStore &keystore, const CScript& scriptPubKey, bool&
         break;
     }
 
+    case TX_HTLC:
+    {
+        // Only consider HTLC's "mine" if we own ALL the keys
+        // involved.
+        vector<valtype> keys;
```
Comment (arielgabizon) [inline src/script/ismine.cpp]: had to add std:: to get it to compile

**Comment by ABISprotocol** [comment]
@ebfull Thanks for mentioning... that was my question, regarding BIP 199, when reading this thread.  

A few questions actually - 
1) Does BIP 199 need to be at final before this (HTLC) can be implemented in Core wallet? Probably dumb process question.
2) Bitcoin wiki on Lightning network states in part, regarding "key features," that "payments can be routed across more than one blockchain (including altcoins and sidechains) as long as all the chains support the same hash function to use for the hash lock, as well as the ability the ability to create time locks."  Although I see that this is not actually part of the proposed functionality here by way of the commits, might it ultimately allow the payment routing across more than one blockchain with further development?
3) TX_HTLC transactions - What is the way in which they are accepted and what would be a sample condition in which one would obviously be rejected?

File: src/policy/policy.cpp (hunk 3 above)
Comment (ABISprotocol) [inline src/policy/policy.cpp]: @arielgabizon Would this actually reject all TX_HTLC items?  I had the impression or understand that there was a design such that these would actually be completed in most instances.  In lnd, a description of this was given as follows:  "If payment/htlc amount is too small, than such output is called dust. If for some reason channel have been force closed during payment, and dust htlc/payment haven't been settled (both sides haven't agreed to remove the htlc and change the balances accordingly), th(e)n we (do) not include it in commitment transaction (thereby giv(ing it) as a fee to miners) in order to make it valid."  For reference to this comment (from a contributor in lnd) see:  https://github.com/lightningnetwork/lnd/pull/115#issuecomment-305739226

With that said, I'm not sure of how the differences will play out between how bitcoin core is planning on handling it vs. how it is handled presently in the lnd development process.  

cc @ebfull

**Comment by jb55** [comment]
This still alive?

**Comment by ebfull** [comment]
Unfortunately I'm too busy to carry the torch on this PR. :(

**Comment by fanquake** [comment]
I'll close this then. If someone wants to restart/pickup this work they can cherry pick/reference and open a new PR.

**Comment by jb55** [comment]
fwiw I've rebased this at https://github.com/jb55/bitcoin/commits/htlc, I'll try to revive it. Should be a fun learning exercise... Will open a new PR if I succeed.

File: qa/rpc-tests/htlc.py [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+        self.nodes = start_nodes(self.num_nodes, self.options.tmpdir)
+        connect_nodes_bi(self.nodes, 0, 1)
+        self.is_network_split = False
+        self.sync_all()
+
+    def activateCSV(self):
+        # activation should happen at block height 432 (3 periods)
+        min_activation_height = 432
+        height = self.nodes[0].getblockcount()
+        assert(height < 432)
```
Comment (takahser) [inline qa/rpc-tests/htlc.py]: why not using `assert(height < min_activation_height)` instead?

Comment (winar-jin) [reply qa/rpc-tests/htlc.py]: I think we can.

**Pull request closed without merging**